internal/deb/extract.go

```diff
@@ -228,22 +227,50 @@ func extractData(dataReader io.Reader, options *ExtractOptions) error {
 			if contentIsCached {
 				pathReader = bytes.NewReader(contentCache)
 			}
-			var targetPath string
+			var relPath string
 			if globPath == "" {
-				targetPath = filepath.Join(options.TargetDir, extractInfo.Path)
+				relPath = extractInfo.Path
 			} else {
-				targetPath = filepath.Join(options.TargetDir, sourcePath)
+				relPath = sourcePath
 			}
 			if extractInfo.Mode != 0 {
 				tarHeader.Mode = int64(extractInfo.Mode)
 			}
-			err := options.Create(&fsutil.CreateOptions{
-				Path:        targetPath,
+			// Create the parent directories using the permissions from the tarball.
+			parents := orderedParentDirs(relPath)
+			for _, path := range parents {
+				if path == "/" {
+					continue
+				}
+				perm, ok := tarDirPermissions[path]
+				if !ok {
+					continue
+				}
+				absPath := filepath.Join(options.TargetDir, path)
+				if _, err := os.Stat(absPath); !os.IsNotExist(err) {
+					// We dont want to this implicit parent directory to overwrite
+					// an existing directory.
```

Comment:
This will indeed optimize the common case you mention where we call `os.Stat` for the same directory several times. I have changed the code to remove the entry in `tarDirPermissions`.

Reply:
What about the first half of the comment: create and check the result, instead of stating and then acting?

Reply:
I am not sure I am understanding precisely what you mean. If we tried to create the folder using `fsutil.Create`, we would create and then check if it exists. When the later happens it overwrites the permissions ([pointer](https://github.com/canonical/chisel/blob/main/internal/fsutil/create.go#L76-L78)). We could change the create API to have a flag for overriding existing fs entries instead of stating here.

Reply:
Right, Create does _mkdir+stat_, precisely because _mkdir_ won't change the mode. Now we're sort of "regretting" and doing _stat+mkdir+stat_ to workaround the fact we're changing the behavior of mkdir. We can simply add e.g. _FixMode_ and make the behavior conditional, so it's _mkdir_ and that's it.

Reply:
Done